Basic Troubleshooting (#126)
PaperMC/docs: docs/paper/admin/how-to/basic-troubleshooting.md

---
slug: /basic-troubleshooting
---

# Basic Troubleshooting

> This guide will help you diagnose your server's problem before reporting it to PaperMC or the plugin's author. Please, read this documentation carefully.

Reviewer: > Please, read this documentation carefully.

Sounds weird, could be omitted.

:::tip Backup Your Server

> People can break their servers while trying to follow this guide. It is highly advised to backup your server before following this guide.  

Reviewer: “Some steps in this guide may cause issues down the road if performed on your production server. It is highly….”

Maybe sounds better?

It would be ideal to create a test server by copying your production server's file, but that's not always possible.  
Please, take a backup first.

:::

> :::caution Stop Your Server

Reviewer: I’m not sure if this is necessary. Stopping the server is somewhat implied by the previous note and things like removing plugins isn’t possible whilst the server is running.

Author: I think mentioning both "stop server" and "backup your server" is needed because I've seen few people trying to take a backup while server is alive in `#paper-help` before. This might look dumb, but it is worth mentioning it.

> Before following this guide, stop your server first. Modifying server files while it is still running will corrupt them. This is like having an operation without anesthesia. Your server will scream like you would.

Reviewer: Funny, but not sure if appropriate.

:::

## Read The Error Message

If your server encounters a problem, it will either print an error message on the server console, create a crash report and close itself, or do both.

If your server crashes, the crash report will be saved in the `crash-report` directory.  

> If your server didn't crash, those error messages will be stored in the `log` directory along with other messages.

Reviewer: The error messages will be stored as part of the logs in the “logs” directory. Note: the logs older than latest will be compressed and not stored as a direct text file.

The first thing you have to do is diagnose those messages.

### Case 01: Paper Watchdog Thread Dump

If your error message looks like this, **do not blindly report it to PaperMC** as it says:

```plaintext
[00:00:00] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH - git-Paper-366 (MC: 1.19.3) ---
[00:00:00] [Paper Watchdog Thread/ERROR]: The server has not responded for 10 seconds! Creating thread dump
[00:00:00] [Paper Watchdog Thread/ERROR]: ------------------------------
```

> This can be caused by various things. Maybe your server's hardware is not enough to run Minecraft server. Maybe one of the plugins you use is causing lag on your server.

Reviewer: “Maybe your servers hardware is not capable of running a Minecraft server”

The thread dump from Paper Watchdog can be found below. If you find any plugin's name in there, talk to the plugin's author about that, not PaperMC.

### Case 02: Stack Trace

> Almost every problem you encounter will print error message lines, which are called "stack trace", on the server console. Examining the stack trace will help you find out what is causing problems on your server.

Reviewer: “… will print an error message, known as a “stack trace”, in the server console”

Author: I consider "error message" + "exception type" + "exception message" + "lines beginning with `at`" to be a single stack trace, so I wrote it that way. If I'm wrong, I will change that sentence.

The stack trace starts with the error message, exception type, and exception message.

> Both error messages and exception messages were put there by the developer of either your plugin or paper server. These messages tell you what problem your server experienced.  

Reviewer: Are these not the same things? Also, I would say that I do not intentionally put exceptions in my code :)

Author: I will try to explain about those words with this stack trace taken from somewhere of `#paper-help`.

``` plaintext
[00:00:00] [Server thread/ERROR]: Could not load 'plugins/commandnpc-2.1.2.jar' in folder 'plugins'
org.bukkit.plugin.UnknownDependencyException: Unknown/missing dependency plugins: [Citizens]. Please download and install these plugins to run 'CommandNPC'.
    at org.bukkit.plugin.SimplePluginManager.loadPlugins(SimplePluginManager.java:289) ~[paper-api-1.19.3-R0.1-SNAPSHOT.jar:?]
    at org.bukkit.craftbukkit.v1_19_R2.CraftServer.loadPlugins(CraftServer.java:436) ~[paper-1.19.3.jar:git-Paper-366]
```

"error message" : `Could not load 'plugins/commandnpc-2.1.2.jar' in folder 'plugins'`  
"exception type" : `org.bukkit.plugin.UnknownDependencyException`  
"exception message" : `Unknown/missing dependency plugins: [Citizens]. Please download and install these plugins to run 'CommandNPC'.`
"body of stack trace" : `at org.bukkit.plugin.SimplePluginManager.loadPlugins(SimplePluginManager.java:289) ~[paper-api-1.19.3-R0.1-SNAPSHOT.jar:?]`, `at org.bukkit.craftbukkit.v1_19_R2.CraftServer.loadPlugins(CraftServer.java:436) ~[paper-1.19.3.jar:git-Paper-366]` ...

And yes, developer of plugin can choose not to include "exception message" in his plugin code. If then, if I remember correctly, "exception message" part will just not displayed in stack trace. That's why I wrote like that.

I could not think of better words for these words like "error message", "exception message" etc.

An exception type like `java.lang.RuntimeException` tells you the type of the exception. This will help the developer (and you) understand the type of problem.

> Many lines beginning with `at` may appear beneath the exception message. These are the bodies of the stack trace. These lines tell you where the problem starts. The top line of the body of the stack trace will tell you exactly where the problem occurred, as well as the name of the JAR file.

Reviewer: “This is the body of the stack trace”

Also not sure if it included the name of the jar file, can’t remember off the top of my head.

<details>
  <summary>Example 01: Server attempted to load chunk saved with newer version of minecraft!</summary>

```plaintext
[00:00:00 WARN]: java.lang.RuntimeException: Server attempted to load chunk saved with newer version of minecraft! 3218 > 3120
```

You tried to load the world generated with a higher version of Minecraft. You cannot do this.  
If you don't have any backup of your world before the chunk version update, you must use your updated world with a higher version of Minecraft.

</details>

<!-- Another examples -->

## Find Out What Plugin Is Causing The Problem

If you can't find the name of any plugin in the thread dump or stack trace, try these steps.

### Disable All Plugins

To determine if it is a plugin or the paper itself that is causing problems, disable all of your plugins first.

You can disable all of your plugins by renaming the `plugins` directory to something else, such as `plugins-disabled`, or by archiving the `plugins` directory and deleting it.

### Run Server Without Any Plugins

Try to run your server without any plugins. If you rename or delete `plugins` directory, your server will run without any plugins.

> If you don't have any issues with this condition, it's the plugin that's causing them.  

Reviewer: “If after removing the plugins, the issue is resolved, you know that it is a plugin causing the issue.”

> If you still experience problems, head to [Paper Documentation](#paper-documentation) because it might be paper that is creating problems, and most of them are because of a misconfigured paper server.

Reviewer: “because it may be that your Paper server is mis-configured and this is creating issues. “

### Binary Search

To find out what plugin is causing problem, [binary search](https://en.wikipedia.org/wiki/Binary_search_algorithm) is ideal way to do it.

Here is brief animation about binary search.

> ![Binary Search Example](assets/binary-search.gif)

Reviewer: I like the animation, but there is a fairly long pause before it starts, can this be lowered?

## Check Plugin Updates

There is a chance that your problem is already fixed in the latest release or latest build of the plugin.

> Head to your plugin's official download page and check if you are using the latest build or the latest release of the plugin. If not, update it to the latest version and try to run your server again to see if the problem is fixed or not.

Reviewer: “to see if the problem is resolved.”

### Update Library Plugin Too

> Many plugins use library plugins, and you have to download them and put them in `plugins` directory.

Reviewer: “… For example ProtocolLib”

If you don't update them to the latest version or latest build, you might experience problems related to plugins that use the library plugin.

> Some library plugins, like ProtocolLib, tell their users to use their latest development build for support of the latest Minecraft version. You should look carefully at the requirements of your plugin.

Reviewer: If you remove the reference to ProtocolLib from here as i added it above. I did this so that it is immediately clear what a library plugin is at the start, rather than realise here.

## Check Documentation

If you misconfigured your plugin or your server, it can also cause problems on your server.

### Plugin Documentation

Many plugins provide their own documentation about how to set them up properly. Read those documents carefully and check if there is something wrong with the plugin's configuration.

### Paper Documentation

Paper can also be configured in a variety of ways. Check these documents for detailed explanations about each configuration.

* [Paper Global Config](../reference/global-configuration.md)
* [Paper Per World Configuration](../reference/world-configuration.md)

## Consult With Developer

> If your problem is related to a plugin you use and you still don't know how to solve it, you can try to reach the plugin's author.  

Reviewer: “… try to reach out to the plugin’s author.”

---
slug: /basic-troubleshooting
---

# Basic Troubleshooting

This guide will help you diagnose your server's problem before reporting it to PaperMC or the plugin's author.

:::caution Stop Your Server And Take A Backup

Before following this guide, stop your server first. Modifying server files while it is still running will corrupt them.  
Only a full server shutdown can prevent this.

Also, if you don't follow this guide carefully or make a mistake while following this guide, you might corrupt your server. It is highly advised to backup your server before following this guide.  
It would be ideal to create a test server by copying your production server's file, but that's not always possible.

> Please, take a backup first.

Reviewer 2: Remove this line

Reviewer: It is just repetitive

Reviewer 2: The heading and text already tell you to create a backup two times, that last line is unnecessary

Author: If I remove whole line according to the suggestion, things would be like:

Heading: Stop and backup your server  
Text: Stop your server

So two times of 'stop your server' and only one 'backup your server'. It looks like something about 'backup your server' is missing in text body. So there should be at least one line about saying 'backup your server'.

Author: Oh, I misunderstood how GitHub's review looks like. So Kenny was taking about the line `Please, take a backup first.`, right?

[00:00:00 WARN]: java.lang.RuntimeException: Server attempted to load chunk saved with newer version of minecraft! 3218 > 3120
```

You tried to load the world generated with a higher version of Minecraft. You cannot do this.  
If you don't have any backup of your world before the chunk version update, you must use your updated world with a higher version of Minecraft.

</details>

<!-- Another examples -->

## Find Out What Plugin Is Causing The Problem

If you can't find the name of any plugin in the thread dump or stack trace, try these steps.

### Disable All Plugins

To determine if it is a plugin or the paper itself that is causing problems, disable all of your plugins first.

You can disable all of your plugins by renaming the `plugins` directory to something else, such as `plugins-disabled`, or by archiving the `plugins` directory and deleting it.

After that, try to run your server.

If the problem is resolved after removing the plugins, you know that it was a plugin that caused the issue.  
If you still experience problems, head to [Paper Documentation](#paper-documentation). Maybe your paper server is misconfigured, and that is creating issues.

### Binary Search

> To find out what plugin is causing problem, [binary search](https://en.wikipedia.org/wiki/Binary_search_algorithm) is ideal way to do it.
> 
> <!-- Animated GIF file -->
> 
> And here is brief explanation about how to do it.
> 

Reviewer 2: ```suggestion
To efficiently search for the plugin that is causing the problem, you can do the following:
```

Author: What about the title, `### Binary Search`? Do you think it is enough or should this also be changed according to the text below of it?

If you don't update them to the latest version or latest build, you might experience problems related to plugins that use the library plugin.

Some library plugins tell their users to use their latest development build for support of the latest Minecraft version. You should look carefully at the requirements of your plugin.

## Check Documentation

If you misconfigured your plugin or your server, it can also cause problems on your server.

### Plugin Documentation

Many plugins provide their own documentation about how to set them up properly. Read those documents carefully and check if there is something wrong with the plugin's configuration.

### Paper Documentation

Paper can also be configured in a variety of ways. Check these documents for detailed explanations about each configuration.

* [Paper Global Config](../reference/global-configuration.md)
* [Paper Per World Configuration](../reference/world-configuration.md)

## Consult With Developer

If your problem is related to a plugin you use and you still don't know how to solve it, you can try to reach out to the plugin's author.  

> Many plugins have a way to contact their author, like GitHub Issue Trakcer, Discord Support Server, Gitter, IRC, etc.  

Reviewer 2: ```suggestion
Many plugins have a way to contact their author, like a GitHub issue tracker, Discord support guild, Gitter, IRC, etc.  
```

> Don't ask for help related to a plugin in PaperMC.

Reviewer 2: Remove this line

> If your problem isn't related to any plugin, you can come to PaperMC's Discord server and ask for help, or create a new issue in the GitHub Issue Tracker.

Reviewer 2: ```suggestion
If your problem isn't related to any plugin, you can come to PaperMC's Discord server and ask for help, or create a new issue on our GitHub issue tracker.
```
